test/harness/services/gossip/adapter/transport_contract_test.go

```diff
@@ -66,28 +72,55 @@ func aTamperingTransport() *transportContractContext {
 	return res
 }
 
-func aMemberlistTransport() *transportContractContext {
-	logger := log.GetLogger()
+func createContractTestConfig(publicKey primitives.Ed25519PublicKey, federationNodes map[string]config.FederationNode) adapter.Config {
+	cfg := config.EmptyConfig()
+	cfg.SetNodePublicKey(publicKey)
+	cfg.SetFederationNodes(federationNodes)
+	cfg.SetDuration(config.GOSSIP_CONNECTION_KEEP_ALIVE_INTERVAL, 20*time.Millisecond)
+	cfg.SetDuration(config.GOSSIP_NETWORK_TIMEOUT, 1*time.Second)
+	return cfg
+}
+
+func aDirectTransport(ctx context.Context) *transportContractContext {
 	res := &transportContractContext{}
-	res.publicKeys = []primitives.Ed25519PublicKey{{0x01}, {0x02}, {0x03}, {0x04}}
-	configs := []adapter.MemberlistGossipConfig{
-		{res.publicKeys[0], 60001, []string{"127.0.0.1:60002", "127.0.0.1:60003", "127.0.0.1:60004"}},
-		{res.publicKeys[1], 60002, []string{"127.0.0.1:60001", "127.0.0.1:60003", "127.0.0.1:60004"}},
-		{res.publicKeys[2], 60003, []string{"127.0.0.1:60001", "127.0.0.1:60002", "127.0.0.1:60004"}},
-		{res.publicKeys[3], 60004, []string{"127.0.0.1:60001", "127.0.0.1:60002", "127.0.0.1:60003"}},
+
+	// randomize listen port between tests to reduce flakiness and chances of listening clashes
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	firstRandomPort := 20000 + r.Intn(40000)
+
+	federationNodes := make(map[string]config.FederationNode)
+	for i := 0; i < 4; i++ {
+		publicKey := keys.Ed25519KeyPairForTests(i).PublicKey()
+		federationNodes[publicKey.KeyForMap()] = config.NewHardCodedFederationNode(publicKey, uint16(firstRandomPort+i), "127.0.0.1")
+		res.publicKeys = append(res.publicKeys, publicKey)
 	}
+
+	configs := []adapter.Config{
+		createContractTestConfig(res.publicKeys[0], federationNodes),
+		createContractTestConfig(res.publicKeys[1], federationNodes),
+		createContractTestConfig(res.publicKeys[2], federationNodes),
+		createContractTestConfig(res.publicKeys[3], federationNodes),
+	}
+
+	logger := log.GetLogger().WithOutput(log.NewOutput(os.Stdout).WithFormatter(log.NewHumanReadableFormatter()))
+
 	res.transports = []adapter.Transport{
-		adapter.NewMemberlistTransport(configs[0], logger),
-		adapter.NewMemberlistTransport(configs[1], logger),
-		adapter.NewMemberlistTransport(configs[2], logger),
-		adapter.NewMemberlistTransport(configs[3], logger),
+		adapter.NewDirectTransport(ctx, configs[0], logger),
+		adapter.NewDirectTransport(ctx, configs[1], logger),
+		adapter.NewDirectTransport(ctx, configs[2], logger),
+		adapter.NewDirectTransport(ctx, configs[3], logger),
 	}
 	res.listeners = []*mockListener{
 		listenTo(res.transports[0], res.publicKeys[0]),
 		listenTo(res.transports[1], res.publicKeys[1]),
 		listenTo(res.transports[2], res.publicKeys[2]),
 		listenTo(res.transports[3], res.publicKeys[3]),
 	}
+
+	// TODO: improve this, we need some time until everybody connects to everybody else
+	// TODO: maybe add an adapter function to check how many active outgoing connections we have
```

Comment:
Adapter could take a `ConnectionListener` that gets notified on connects/disconnects, and the test could provide such a listener to block until the desired number of connections has been reached

Reply:
Added your proposal to the TODO, will resolve outside this PR